### DISCUSSION TITLE
HTTPError: 401 Client Error: Unauthorized for url: https://api.openai.com/v1/completions

### DISCUSSION THREAD

**Discussion by TouchstoneTheDev** [TheR1D/shell_gpt#33] [Q&A] [answered]
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /home/c/.local/lib/python3.10/site-packages/sgpt.py:150 in main                                  │
│                                                                                                  │
│   147 │   │   max_tokens = 1024                                                                  │
│   148 │   if editor:                                                                             │
│   149 │   │   prompt = get_edited_prompt()                                                       │
│ ❱ 150 │   response_text = openai_request(prompt, model, max_tokens, api_key, temperature, top_   │
│   151 │   # For some reason OpenAI returns several leading/trailing white spaces.                │
│   152 │   response_text = response_text.strip()                                                  │
│   153 │   typer_writer(response_text, code, shell, animation)                                    │
│                                                                                                  │
│ ╭─────────────────────────────────────────── locals ───────────────────────────────────────────╮ │
│ │       animation = True                                                                       │ │
│ │         api_key = :)   │ │
│ │            code = False                                                                      │ │
│ │          editor = False                                                                      │ │
│ │         execute = False                                                                      │ │
│ │      max_tokens = 2048                                                                       │ │
│ │           model = <Model.davinci: 'text-davinci-003'>                                        │ │
│ │          prompt = 'nginx default config file location'                                       │ │
│ │           shell = False                                                                      │ │
│ │         spinner = True                                                                       │ │
│ │     temperature = 1.0                                                                        │ │
│ │ top_probability = 1.0                                                                        │ │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────╯ │
│                                                                                                  │
│ /home/c/.local/lib/python3.10/site-packages/sgpt.py:69 in wrapper                                │
│                                                                                                  │
│    66 │   │   text = TextColumn("[green]Requesting OpenAI...")                                   │
│    67 │   │   with Progress(SpinnerColumn(), text, transient=True) as progress:                  │
│    68 │   │   │   progress.add_task("request")                                                   │
│ ❱  69 │   │   │   return func(*args, **kwargs)                                                   │
│    70 │                                                                                          │
│    71 │   return wrapper                                                                         │
│    72                                                                                            │
│                                                                                                  │
│ ╭─────────────────────────────────────────── locals ───────────────────────────────────────────╮ │
│ │     args = (                                                                                 │ │
│ │            │   'nginx default config file location',                                         │ │
│ │            │   <Model.davinci: 'text-davinci-003'>,                                          │ │
│ │            │   2048,                                                                         │ │
│ │            │                                                                                 │ │
│ │            '   :)
 │ │
│ │            │   1.0,                                                                          │ │
│ │            │   1.0                                                                           │ │
│ │            )                                                                                 │ │
│ │     func = <function openai_request at 0x7f39b3a579a0>                                       │ │
│ │   kwargs = {}                                                                                │ │
│ │ progress = <rich.progress.Progress object at 0x7f39b3a32fe0>                                 │ │
│ │     text = <rich.progress.TextColumn object at 0x7f39b3a33370>                               │ │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────╯ │
│                                                                                                  │
│ /home/c/.local/lib/python3.10/site-packages/sgpt.py:101 in openai_request                        │
│                                                                                                  │
│    98 │   │   "top_p": top_p,                                                                    │
│    99 │   }                                                                                      │
│   100 │   response = requests.post(API_URL, headers=headers, json=data, timeout=180)             │
│ ❱ 101 │   response.raise_for_status()                                                            │
│   102 │   return response.json()["choices"][0]["text"]                                           │
│   103                                                                                            │
│   104                                                                                            │
│                                                                                                  │
│ ╭─────────────────────────────────────────── locals ───────────────────────────────────────────╮ │
│ │     api_key = '  :)  │ │
│ │        data = {                                                                              │ │
│ │               │   'prompt': 'nginx default config file location',                            │ │
│ │               │   'model': <Model.davinci: 'text-davinci-003'>,                              │ │
│ │               │   'max_tokens': 2048,                                                        │ │
│ │               │   'temperature': 1.0,                                                        │ │
│ │               │   'top_p': 1.0                                                               │ │
│ │               }                                                                              │ │
│ │     headers = {                                                                              │ │
│ │               │   'Content-Type': 'application/json',                                        │ │
│ │               │   'Authorization': 'Bearer                                                   │ │
│ │              :)  
│ │
│ │               }                                                                              │ │
│ │  max_tokens = 2048                                                                           │ │
│ │       model = <Model.davinci: 'text-davinci-003'>                                            │ │
│ │      prompt = 'nginx default config file location'                                           │ │
│ │    response = <Response [401]>                                                               │ │
│ │ temperature = 1.0                                                                            │ │
│ │       top_p = 1.0                                                                            │ │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────╯ │
│                                                                                                  │
│ /home/c/.local/lib/python3.10/site-packages/requests/models.py:1021 in raise_for_status          │
│                                                                                                  │
│   1018 │   │   │   )                                                                             │
│   1019 │   │                                                                                     │
│   1020 │   │   if http_error_msg:                                                                │
│ ❱ 1021 │   │   │   raise HTTPError(http_error_msg, response=self)                                │
│   1022 │                                                                                         │
│   1023 │   def close(self):                                                                      │
│   1024 │   │   """Releases the connection back to the pool. Once this method has been            │
│                                                                                                  │
│ ╭─────────────────────────────────────────── locals ───────────────────────────────────────────╮ │
│ │ http_error_msg = '401 Client Error: Unauthorized for url:                                    │ │
│ │                  https://api.openai.com/v1/completions'                                      │ │
│ │         reason = 'Unauthorized'                                                              │ │
│ │           self = <Response [401]>                                                            │ │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────╯ │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯

**Comment by TheR1D** [comment]
This happens since you provided ":)" as your API key. Follow [these](https://github.com/TheR1D/shell_gpt/issues/5#issuecomment-1422754392) steps to change your API key.

**Reply by TouchstoneTheDev** [reply]
I don't want to expose my keys so i just changed it in this post to :)

**Reply by TouchstoneTheDev** [reply]
btw thenks for above link/reference will try

**Reply by TheR1D** [reply]
Since you didn't mention it, I thought this is a real value 😀. Just make sure you have right API key in `~/.config/shell-gpt/api_key.txt`

**Comment by TouchstoneTheDev** [comment]
Now I cannot access chatgpt
it's just checking my browser
cannot go further with it !

**Comment by ghost** [comment, marked as answer]
Hiya, I've ran into the same problem, and maybe I can help with the situation?

I had to make sure one I messed this up that I removed all traces of Shell GPT. 
I removed all related packages from (they are labeled accordingly)
>```/Users/<user>/Library/Python/3.11/lib/python/site-packages```
^ I also noticed and api_key.txt here, but I think its been removed..


Then, I noticed after a clean installation, that the same api string was still being used. After some looking around, I found my .config file.. You must either edit this, or completely delete it and start fresh :) 

*If* you would like to remove it, use ```ls -a```, and open your .config directory. It is a hidden directory. From here you will see the dir ```shell-gpt/```. Then you've found the right file! ```rm api_key.txt```
***Since you are starting fresh*** Please think about creating a new api key, just to eliminate anymore confusion.

*Otherwise* ... this will also work ```~/.config/shell-gpt/api_key.txt```, but just incase you'd like to know where everything is at, please use the prior. Either way, edit this file once more! :D

If you're still having anymore trouble, double check your path extension. I notice that it might not automatically be added everytime, but it's seriously such an easy fix. The extension can be added by using the following:
>```export "/Users/<user>/Library/Python/3.11/bin/:$PATH"```
> ```echo "PATH=$PATH:/Users/<user>/Library/Python/3.11/bin/" >> "$HOME/.$(basename $SHELL)rc"```

The later choice might be your best option. After I did this, it worked. Every instance in one terminal, in different shells, in two sperate terminals, etc. I have noticed a 500 Error, but these, and other related issues are because of the OpenAI API. 

*Kudos* to the creator for this. I can imagine how overwhelming these can get, so hopefully this helps @tanmay-veer?

**Comment by TouchstoneTheDev** [comment]
Ig no problem now using shellg bur I am not able to access chatgpt :()

**Comment by TouchstoneTheDev** [comment]
is there any guide for prompt engineering or doc for text-davinci-003

**Comment by ghost** [comment]
@tanmay-veer yeah, there might be more around the internet refer to this [website](https://platform.openai.com/docs/api-reference/completions) for the help with completions, and refer to this [website](https://platform.openai.com/docs/api-reference/models/list) for models. I would recommend still using davinci because of its token cost.
